### PR TITLE
Allow the use of toga.Window as the main window

### LINKED ISSUE #1870: Add a Toga.SimpleApp base class
### What is the problem or limitation you are having?

Toga.App includes a number of features that may not be desirable on all apps. 

On Windows and GTK, the main window of a Toga app includes a menubar.

On iOS and Android, the main window contains a titlebar.

The titlebar provided by the web backend may also be undesirable in some situations.

### Describe the solution you'd like

Toga should offer a `Toga.SimpleApp` base class that only includes the bare minimum of decorations.

`SimpleApp` should not support menus or toolbars. It should not draw a titlebar on its main window. Any commands that are registered should only be accessible as keyboard shortcuts.

`Toga.App` should subclass `Toga.SimpleApp` to add these additional features.

The MainWindow instantiated by the app may also need a "Simple" variant.

### Describe alternatives you've considered

Provide a formal API to turn off menus, toolbars etc on a feature-by-feature basis. This is problematic because of platforms like macOS, where menus are non-optional.

### Additional context

#733 was an original request for this feature.

### PR DESCRIPTION
**Pull request by freakboy3742**
Builds on #2646; that PR should be reviewed and merged before reviewing this one.

Derived from #2244.

Adds the ability to use `toga.Window` as the main window of an app. As `toga.Window` cannot have a toolbar, this is a good indicator that the user wants a minimalist "Simple" app, so the set of menu commands that are installed is also reduced. A SimpleApp example has been added to demonstrate this capability.

We need to know the type of the main window before we create the app commands; this means the app commands must be created *after* `startup()` has been invoked. However, app commands are currently created *before* startup so that the `startup()` method can customise the default commands. 

To allow for customisation, `create_app_commands()` has been promoted to a public interface method so that users can override `create_app_commands()` in their apps; but this method is invoke *after* `startup()` so that it has sufficient context to create the right set of default commands.

Fixes #1870.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by mhsmith** [comment]
> As `toga.Window` cannot have a toolbar, this is a good indicator that the user wants a minimalist "Simple" app, so the set of menu commands that are installed is also reduced.

On macOS, which is the platform where this makes the greatest difference, doesn't this remove commands which are needed for basic functionality like window management UI (Hide, Minimize, etc.), and text editing (Cut, Paste, etc.)?

More generally, it looks to me like the existing set of commands is already minimalist enough, so introducing these two variants seems to add a lot of complexity for no clear benefit.

**Comment by freakboy3742** [comment]
> > As `toga.Window` cannot have a toolbar, this is a good indicator that the user wants a minimalist "Simple" app, so the set of menu commands that are installed is also reduced.
> 
> On macOS, which is the platform where this makes the greatest difference, doesn't this remove commands which are needed for basic functionality like window management (Hide, Minimize, etc.), and text editing (Cut, Paste, etc.)?
> 
> More generally, it looks to me like the existing set of commands is already minimalist enough, so introducing these two variants seems to add a lot of complexity for no clear benefit.

My original thinking was that a minimal app should be truly as minimal as possible - but you make a good point about Hide, Minimise, Cut, Paste et al. Looking over the list, the only ones that could be considered *completely* optional are Preferences and Visit Homepage. Visit homepage should arguably exist by default, as the Apple HIG suggests apps [should always provide help](https://developer.apple.com/design/human-interface-guidelines/the-menu-bar#Help-menu); "visit homepage" isn't the best manifestation of help, but we need *something* to provide the default help menu that has the "search" option to find menu items.

We will still need the "minimal vs standard" API for other platforms, because we don't want to create command instances that literally won't (or can't) be used. 

I think we can roll this into the description of MainWindow - pointing out that if the platform uses app-level menus, using a MainWindow as the main window may add extra menu items. That then means we can hide the "preferences" item from bare bones apps, and provides a lead in for the introduction of DocumentMainWindow, which will introduce document management commands (new, open, save etc).

> Also, the `main_window` documentation is about to become a lot more complicated in the next PR, which adds "background" and "session" apps. This would be helped if we eliminate the idea of a "simple app", and make "simple" or "main" purely a property of a window.

Agreed - this is a distinction that matters for internal reasons, but doesn't really matter from the perspective of the external API surface. I've simplified the docs to reflect this.

**Comment by mhsmith** [comment]
> We will still need the "minimal vs standard" API for other platforms, because we don't want to create command instances that literally won't (or can't) be used.

Even if a platform shows no menu bar with a base class `Window`, the commands could still be used if the app creates a `MainWindow` later. For example. when Android Studio has no projects open, it shows a window with a list of recent projects. On Windows, this window has no menu bar, but the per-project windows that it opens do.

> I think we can roll this into the description of MainWindow - pointing out that if the platform uses app-level menus, using a MainWindow as the main window may add extra menu items. That then means we can hide the "preferences" item from bare bones apps, and provides a lead in for the introduction of DocumentMainWindow, which will introduce document management commands (new, open, save etc).

This still seems like a needless coupling between the app and the initial `main_window`:

* It permanently commits the app to a set of commands based on what its initial `main_window` happens to be, even though `main_window` can be reassigned later. 
 
* It adds clutter to the App class by adding a new method which will rarely be overridden, while also adding complexity because it *must* be overridden to remove default commands, which you would naturally expect to be able to do in `startup` at the same time as you create your own commands.

* I don't see how it generalizes to `main_window=None`, because we would have no idea whether the app's windows are going to be MainWindows or not.

I think all the goals of this change can be achieved in other, more flexible ways:

* In the case of Preferences, the command can appear or disappear based on whether `App.preferences` has been overridden. Why should an app that doesn't have a preferences command be forced to clutter its user experience with a disabled menu item?

* If DocumentMainWindow is going to have document management commands associated with it, that could be done with #2210, which is context-sensitive, not permanent.

**Comment by freakboy3742** [comment]
As discussed in person:
* I've merged the "minimal" and "standard" app commands back into `create_app_commands`, which means we can go back to auto-creating the app commands before `startup` is invoked
* The default commands are *always* installed, regardless of the main window type, and whether platform displays app-level menus..
* I've modified the "Preferences" menu item to only be installed if the user overrides the `preferences()` method.

File: docs/reference/api/app.rst [hunk 1]
```diff
@@ -72,11 +72,34 @@ format compatible with :any:`importlib.metadata`. If you deploy your app with `B
...
+
+This is the type of app that will be created if you use an instance of :any:`toga.App`
+passing in a ``startup`` argument to the constructor.
 
 Notes
 -----
 
+* On macOS, menus are tied to the app, not the window; and a menu is mandatory.
+  Therefore, a minimal macOS app (i.e., an app using a :any:`toga.Window` as the main
+  window) will still have a menu, but it will only have the bare minimum of menu items.
```
Comment (mhsmith) [inline docs/reference/api/app.rst:101]: This paragraph is no longer relevant – a macOS app will show whatever commands are in `App.commands`, regardless of whether you use a Window or a MainWindow.

```suggestion
```

File: changes/2649.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Toga apps now expose a ``create_app_commands()`` method that can be overridden to customize the set of default menu items that are installed.
```
Comment (mhsmith) [inline changes/2649.feature.rst:1]: This change note should be removed.

File: docs/reference/api/mainwindow.rst [hunk 3]
```diff
@@ -55,10 +55,20 @@ Usage
...
-:class:`toga.MainWindow` instance will display a menu bar that contains the app
-control commands (such as About, Quit, and anything else required by the
-platform's HIG). It may also contain a toolbar.
+interface to an application. A :class:`toga.MainWindow` may optionally have a toolbar.
+The presentation of :class:`toga.MainWindow` is platform dependent:
+
+* On desktop platforms that place menus inside windows (e.g., Windows, and most Linux
+  window managers), a :class:`toga.MainWindow` instance will display a menu bar that
+  contains the app control commands (such as About, Quit, and anything else required by
+  the platform's HIG).
```
Comment (mhsmith) [inline docs/reference/api/mainwindow.rst:64]: This makes user-defined commands seem like an afterthought. In fact, I don't think it's helpful to distinguish between HIG commands and user-defined commands here – that can be done on the App or Commands pages.

```suggestion
* On desktop platforms that place menus inside windows (e.g., Windows, and most Linux
  window managers), a :class:`toga.MainWindow` instance will display a menu bar that
  contains :attr:`App.commands`.
```

File: docs/reference/api/mainwindow.rst [hunk 4]
```diff
@@ -55,10 +55,20 @@ Usage
...
+
+* On desktop platforms that use an app-level menu bar (e.g., macOS, and some Linux
+  window managers), the window will not have a menu bar; all menu items will be
+  displayed in the app bar.
+
+* On mobile, web and console platforms, a :class:`toga.MainWindow` will include a title
+  bar that can contain both menus and toolbar items.
 
 In addition to the platform's default commands, user-defined commands can be
 added to the :class:`toga.MainWindow`'s menu by adding them to
```
Comment (mhsmith) [inline docs/reference/api/mainwindow.rst:74]: I can't post a suggestion on an unmodified line, but if the previous suggestion is accepted, this sentence should be removed.

File: winforms/src/toga_winforms/app.py [hunk 5]
```diff
@@ -172,6 +163,18 @@ def create_app_commands(self):
...
         )
 
+        # If the user has overridden preferences, provide a menu item.
+        if overridden(self.interface.preferences):
+            self.interface.commands.add(
+                Command(
+                    simple_handler(self.interface.preferences),
+                    "Preferences",
+                    group=Group.FILE,
+                    enabled=overridden(self.interface.preferences),
```
Comment (mhsmith) [inline winforms/src/toga_winforms/app.py:173]:
```suggestion
```

**Review by mhsmith — COMMENTED** [review]
> I've modified the "Preferences" menu item to only be installed if the user overrides the `preferences()` method.

This should be documented at `App.preferences` and `Command.PREFERENCES`. 

It should also be mentioned in the existing 90.feature.rst change note, which should probably be renamed to `removal`, because if anyone was using the command before, it will disappear until they implement `App.preferences`.

**Review by mhsmith — APPROVED** [review]

**Pull request merged**